gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/MessageConsumerImpl.java

```diff
@@ -0,0 +1,281 @@
+  private final AtomicInteger queuedCallbacks;
+  private final int maxQueuedCallbacks;
+  private final Object futureLock = new Object();
+  private final Runnable scheduleRunnable;
+  private boolean closed;
+  private Future<?> scheduledFuture;
+  private PullFuture pullerFuture;
+  private boolean stopped = true;
+
+  /**
+   * Default executor factory for the message processor executor. By default a single-threaded
+   * executor is used.
+   */
+  static class DefaultExecutorFactory implements ExecutorFactory<ExecutorService> {
+
+    private final ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    @Override
+    public ExecutorService get() {
+      return executor;
+    }
+
+    @Override
+    public void release(ExecutorService executor) {
+      executor.shutdownNow();
+    }
+  }
+
+  private MessageConsumerImpl(Builder builder) {
+    this.pubsubOptions = builder.pubsubOptions;
+    this.subscription = builder.subscription;
+    this.messageProcessor = builder.messageProcessor;
+    this.pubsubRpc = pubsubOptions.rpc();
+    this.pubsub = pubsubOptions.service();
+    this.deadlineRenewer = builder.deadlineRenewer;
+    this.queuedCallbacks = new AtomicInteger();
+    this.timer = SharedResourceHolder.get(TIMER);
+    this.executorFactory = firstNonNull(builder.executorFactory, new DefaultExecutorFactory());
+    this.executor = executorFactory.get();
+    this.maxQueuedCallbacks = firstNonNull(builder.maxQueuedCallbacks, MAX_QUEUED_CALLBACKS);
+    this.scheduleRunnable = new Runnable() {
+      @Override
+      public void run() {
+        synchronized (futureLock) {
+          if (closed) {
+            return;
+          }
+          pull();
+        }
+      }
+    };
+    nextPull();
+  }
+
+  private Runnable ackingRunnable(final ReceivedMessage receivedMessage) {
+    return new Runnable() {
+      @Override
+      public void run() {
+        try {
+          messageProcessor.process(receivedMessage);
+          pubsub.ackAsync(receivedMessage.subscription(), receivedMessage.ackId());
+        } catch (Exception ex) {
+          pubsub.nackAsync(receivedMessage.subscription(), receivedMessage.ackId());
+        } finally {
+          deadlineRenewer.remove(receivedMessage.subscription(), receivedMessage.ackId());
+          queuedCallbacks.decrementAndGet();
+          // We can now pull more messages. We do not pull immediately to possibly wait for other
+          // callbacks to end
+          scheduleNextPull(500, TimeUnit.MILLISECONDS);
+        }
+      }
+    };
+  }
+
+  private PullRequest createPullRequest() {
+    return PullRequest.newBuilder()
+        .setSubscription(formatSubscriptionName(pubsubOptions.projectId(), subscription))
+        .setMaxMessages(maxQueuedCallbacks - queuedCallbacks.get())
+        .setReturnImmediately(false)
+        .build();
+  }
+
+  private void scheduleNextPull(long delay, TimeUnit timeUnit) {
+    synchronized (futureLock) {
+      if (!closed && stopped) {
```

Comment:
I think it would be more clear if you called (or treated) `stopped` as `pendingPull` (could be Boolean or the reference to the future itself). and you can probably have only one that at time is scheduled immediately (at start or after a pull [and before adding to the executor] if pull target was not reached) and sometimes being delayed (after being consumed).

I think right now you have an issue as `scheduleNextPull` is calling `pull` directly and does not go through `nextPull`, so you can potentially have multiple pulls beyond target.